torch_xla/utils/gcsfs.py

```diff
@@ -0,0 +1,199 @@
+    return self._wfile.readlines(hint=hint)
+
+  def writelines(self, lines):
+    return self._wfile.writelines(lines)
+
+  def read(self, size=-1):
+    return self._wfile.read(size=-1)
+
+  def readall(self):
+    return self._wfile.readall()
+
+  def readinto(self, bbuf):
+    return self._wfile.readinto(bbuf)
+
+  def write(self, bbuf):
+    return self._wfile.write(bbuf)
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, type, value, traceback):
+    self.close()
+
+
+def open(path, mode='r', encoding='utf-8'):
+  """Opens a Google Cloud Storage (GCS) file for reading or writing.
+
+  Args:
+    path (string): The GCS path of the file. Must be "gs://BUCKET_NAME/PATH"
+      where ``BUCKET_NAME`` is the name of the GCS bucket, and ``PATH`` is a `/`
+      delimited path.
+    mode (string, optional): The open mode, similar to the ``open()`` API.
+      Default: 'r'
+    encoding (string, optional): The character encoding to be used to decode
+      bytes into strings when opening in text mode.
+      Default: 'utf-8'
+
+  Returns:
+    The GCS file object.
+  """
+  if mode.startswith('w'):
+    return WriteableFile(path)
+  gcs_client = gcs.Client()
+  blob = gcs.Blob.from_string(path, client=gcs_client)
+  if mode.startswith('a') or mode.startswith('r+'):
+    data = blob.download_as_string() if blob.exists() else None
+    return WriteableFile(path, init_data=data, append=mode.startswith('a'))
+  data = blob.download_as_string()
+  if mode.find('t') >= 0:
+    return io.StringIO(data.decode(encoding))
+  return io.BytesIO(data)
+
+
+def _get_blob_path(bpath):
+  # The paths returned by the list_blobs() API have the
+  # '/b/BUCKET_NAME/o/BLOB_PATH' format.
+  m = re.match(r'/b/[^/]+/o/(.+)', bpath)
+  if not m:
+    raise RuntimeError('GCS invalid blob path: {}'.format(bpath))
+  return m.group(1)
+
+
+def list(path):
+  """Lists the content of a GCS bucket.
+
+  Args:
+    path (string): The GCS path of the file. Must be "gs://BUCKET_NAME/PATH"
+      where ``BUCKET_NAME`` is the name of the GCS bucket, and ``PATH`` is a `/`
+      delimited path.
+
+  Returns:
+    A list of ``GcsBlob`` object, having ``path`` and ``size`` fields.
+  """
+  m = re.match(r'gs://([^/]+)(.*)', path)
+  if not m:
+    raise ValueError('GCS invalid path: {}'.format(path))
+  bucket_name = m.group(1)
+  bpath = m.group(2) if len(m.group(2)) > 1 else None
+  gcs_client = gcs.Client()
+  blobs = []
+  for blob in gcs_client.list_blobs(bucket_name, prefix=bpath, delimiter='/'):
+    blobs.append(GcsBlob(path=_get_blob_path(blob.path), size=blob.size))
+  return blobs
+
+
+def remove(path):
+  """Removes a GCS blob.
+
+  Args:
+    path (string): The GCS path of the file. Must be "gs://BUCKET_NAME/PATH"
+      where ``BUCKET_NAME`` is the name of the GCS bucket, and ``PATH`` is a `/`
+      delimited path.
+  """
+  m = re.match(r'gs://([^/]+)/(.+)', path)
+  if not m:
+    raise ValueError('GCS invalid path: {}'.format(path))
```

Comment:
Mention this exception in the `Raises:` section of the docstring

Reply:
SG!